filc/retrieval.go

```diff
@@ -0,0 +1,416 @@
+					log.Debugf("Failed to connect to IPFS provider %s: %v", provider, err)
+					continue
+				}
+				connectedCount++
+			}
+
+			// If we were able to connect to at least one of the providers, go ahead
+			// with the retrieval
+
+			var progressLk sync.Mutex
+			var bytesRetrieved uint64 = 0
+			startTime := time.Now()
+
+			if connectedCount > 0 {
+				log.Infof("Successfully connected to %v/%v providers", connectedCount, len(providers))
+
+				log.Info("Starting IPFS retrieval")
+
+				bserv := blockservice.New(node.Blockstore, node.Bitswap)
+				dserv := merkledag.NewDAGService(bserv)
+				//dsess := dserv.Session(ctx)
+
+				cset := cid.NewSet()
+				if err := merkledag.Walk(ctx, func(ctx context.Context, c cid.Cid) ([]*ipldformat.Link, error) {
+					node, err := dserv.Get(ctx, c)
+					if err != nil {
+						return nil, err
+					}
+
+					progressLk.Lock()
+					nodeSize, err := node.Size()
+					if err != nil {
+						nodeSize = 0
+					}
+					bytesRetrieved += nodeSize
+					fmt.Printf("%v (%v)\r", bytesRetrieved, humanize.IBytes(bytesRetrieved))
+					progressLk.Unlock()
+
+					if c.Type() == cid.Raw {
+						return nil, nil
+					}
+
+					return node.Links(), nil
+				}, c, cset.Visit, merkledag.Concurrent()); err != nil {
+					return nil, err
+				}
+
+				log.Info("IPFS retrieval succeeded")
+
+				return &IPFSRetrievalStats{
+					ByteSize: bytesRetrieved,
+					Duration: time.Since(startTime),
+				}, nil
+			} else {
+				log.Info("Could not connect to any providers, will not attempt IPFS retrieval")
+			}
+		} else {
+			log.Info("Could not find the CID on IPFS")
+		}
+	}
+
+	// If no miners are provided, there's nothing else we can do
+	if len(candidates) == 0 {
+		log.Info("No miners were provided, will not attempt FIL retrieval")
+		return nil, xerrors.Errorf("retrieval failed: no miners were provided")
+	}
+
+	// If IPFS retrieval was unavailable, do a full FIL retrieval. Start with
+	// querying all the candidates for sorting.
+
+	log.Info("Querying FIL retrieval candidates...")
+
+	type CandidateQuery struct {
+		Candidate RetrievalCandidate
+		Response  *retrievalmarket.QueryResponse
+	}
+	checked := 0
+	var queries []CandidateQuery
+	var queriesLk sync.Mutex
+
+	var wg sync.WaitGroup
+	wg.Add(len(candidates))
+
+	for _, candidate := range candidates {
+
+		// Copy into loop, cursed go
+		candidate := candidate
+
+		go func() {
+			defer wg.Done()
+
+			query, err := fc.RetrievalQuery(ctx, candidate.Miner, candidate.RootCid)
+			if err != nil {
+				log.Debugf("Retrieval query for miner %s failed: %v", candidate.Miner, err)
+				return
+			}
+
+			queriesLk.Lock()
+			queries = append(queries, CandidateQuery{Candidate: candidate, Response: query})
+			checked++
+			fmt.Printf("%v/%v\r", checked, len(candidates))
+			queriesLk.Unlock()
+		}()
+	}
+
+	wg.Wait()
+
+	log.Infof("Got back %v retrieval query results of a total of %v candidates", len(queries), len(candidates))
```

Comment:
You can ( and probably should ) do this entire discovery phase concurrently with IPFS bootstraping+provider lookup.

The miner lookup is free fil-wise, so no reason not to do it while spinning up IPFS: much better UX

Reply:
good suggestion, and will get to it in a subsequent pr. i'm planning a reorganization that'll make doing this a lot easier. i have more plans for improving this function as well, but it is just in a bit of a messy state rn.